proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+
+```c#
+ResourceLimiter rateLimiter = new SomeRateLimiter(options => ...)
+
+// Synchronous checks
+endpoints.MapGet("/acquire", context =>
+{
+    // Check limiter using `Acquire` that should complete immediately
+    if (rateLimiter.Acquire().IsAcquired)
+    {
+        // Resource was successfully obtained, process the request
+        return context.Response.WriteAsync("Hello World!");
+    }
+    else
+    {
+        // Resource could not be obtained, send 429 response
+        context.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+        return Task.CompletedTask;
+    }
+}
+
+// Async checks
+endpoints.MapGet("/acquireAsync", async context =>
+{
+    if (await rateLimiter.AcquireAsync().IsAcquired)
+    {
+        // Check limiter using `AcquireAsync` which may complete immediately
+        // or wait until resources are available.
+        await context.Response.WriteAsync("Hello World!");
+    }
+    else
+    {
+        // This exception may be thrown when hard cap is reached
+        context.Response.StatusCode = StatusCodes.Status429TooManyRequests;
+        return;
+    }
+}
+```
+
+For additional usage samples, checkout the [proof of concepts](#proof-of-concepts).
+
+### Concrete implementations
+
+#### Default limiter implementations
+
+Additional public APIs are needed to configure specific limiter implementations listed in [the goals](#resource-limit-implementations). The options for configuring these implementations will be described here without going into implementation details.
+
+```c#
+namespace System.Threading.ResourceLimits
+{
+    // Limiter implementations
+    public sealed class FixedWindowRateLimiter : ResourceLimiter { }
+    public sealed class SlidingWindowRateLimiter : ResourceLimiter { }
+    public sealed class TokenBucketRateLimiter : ResourceLimiter { }
+    public sealed class ConcurrencyLimiter : ResourceLimiter { }
+
+    // This specifies the behaviour of `AcquireAsync` When ResourceLimit has been reached
+    public enum ResourceDepletedMode
+    {
+        RejectIncomingRequest,
+        EnqueueIncomingRequest,
+        PushIncomingRequest
+    }
+
+    // Limiter options common to default runtime implementations
+    internal abstract class ResourceLimiterOptions
+    {
+        // Specifies the maximum number of resources for the limiter
+        public long ResourceLimit { get; set; }
+        // Resource depleted mode, configures `AcquireAsync` behaviour
+        public ResourceDepletedMode ResourceDepletedMode { get; set; }
+        // Queue size when queuing is enabled
+        public long MaxQueueSize { get; set; }
+    }
+
+    // Rate limiter options
+    internal abstract class RateLimiterOptions : ResourceLimiterOptions
+    {
+        // This can be used to externally trigger rate limit window/replenishment updates
+        public EventHandler<DateTimeOffset> TickEventHandler { get; set; }
```

Comment:
I'm not a fan of event handlers. Why do we need this? It's not async friendly in it's current state either...

Reply:
What's a better way to trigger a resource replenishment externally for a rate limiter? I think it's worse to have each limiter create a Timer to trigger self-replenishment but I'm open to ideas.

Reply:
Likely it will involve the implementation having an explicit `Replenish(DateTimeOffset timestamp)` that can be called.